src/parsers/power_system_table_data.jl

```diff
@@ -807,99 +807,160 @@ function make_generator(data::PowerSystemTableData, gen, cost_colnames, bus, gen
-            c in var_cost if !in(nothing, c)
-        ])
-        if isempty(var_cost)
-            @warn "Unable to calculate variable cost for $(gen.name)" var_cost maxlog = 5
-        end
-    else
-        var_cost = [(0.0, 0.0)]
-    end
+    startup_cost, shutdown_cost = calculate_uc_cost(data, gen, fuel_price)
 
-    if length(var_cost) > 1
-        var_cost[2:end] = [
-            (
-                (
-                    var_cost[i][1] * fuel_cost * (var_cost[i][2] - var_cost[i - 1][2]) +
-                    var_cost[i][2] * vom
-                ),
-                var_cost[i][2],
-            ) .* gen.active_power_limits_max .* base_power for i in 2:length(var_cost)
-        ]
-        var_cost[1] =
-            ((var_cost[1][1] * fuel_cost + vom) * var_cost[1][2], var_cost[1][2]) .*
-            gen.active_power_limits_max .* base_power
-
-        fixed = max(
-            0.0,
-            var_cost[1][1] -
-            (var_cost[2][1] / (var_cost[2][2] - var_cost[1][2]) * var_cost[1][2]),
-        )
-        var_cost[1] = (var_cost[1][1] - fixed, var_cost[1][2])
+    op_cost = ThermalGenerationCost(
+        FuelCurve(var_cost, UnitSystem.NATURAL_UNITS, fuel_price),
+        fixed * fuel_price,
+        startup_cost,
+        shutdown_cost,
+    )
+    return op_cost
+end
+
+function make_cost(
+    ::Type{T},
+    data,
+    gen,
+    cost_colnames::_CostPointColumns,
+) where {T <: ThermalGen}
+    cost_pairs = get_cost_pairs(gen, cost_colnames)
+    var_cost = create_pwl_cost(cost_pairs)
+    startup_cost, shutdown_cost = calculate_uc_cost(data, gen, fuel_price)
 
-        for i in 2:length(var_cost)
-            var_cost[i] = (var_cost[i - 1][1] + var_cost[i][1], var_cost[i][2])
+    op_cost = ThermalGenerationCost(
+        CostCurve(var_cost, UnitSystem.NATURAL_UNITS),
+        gen.fixed_cost,
+        startup_cost,
+        shutdown_cost,
+    )
+    return op_cost
+end
+
+function make_cost(
+    ::Type{T},
+    data,
+    gen,
+    cost_colnames::_HeatRateColumns,
+) where {T <: HydroGen}
+    fuel_price = gen.fuel_price / 1000.0
+    cost_pairs = get_cost_pairs(gen, cost_colnames)
+    var_cost, fixed = create_pwinc_cost(cost_pairs)
+    op_cost = HydroGenerationCost(
+        FuelCurve(var_cost, UnitSystem.NATURAL_UNITS, fuel_price),
+        fixed * fuel_price)
+    return op_cost
+end
+
+function make_cost(
+    ::Type{T},
+    data,
+    gen,
+    cost_colnames::_CostPointColumns,
+) where {T <: HydroGen}
+    cost_pairs = get_cost_pairs(gen, cost_colnames)
+    var_cost = create_pwl_cost(cost_pairs)
+    op_cost = HydroGenerationCost(
+        CostCurve(var_cost, UnitSystem.NATURAL_UNITS),
+        gen.fixed_cost)
+    return op_cost
+end
+
+function make_cost(
+    ::Type{T},
+    data,
+    gen,
+    cost_colnames::_HeatRateColumns,
+) where {T <: RenewableGen}
+    @warn "Heat rate parsing not valid for RenewableGen replacing with zero cost"
```

Comment:
I don't want to throw an error because the RTS data has zeros in some of these columns, that would render PSY4 incompatible with the RTS data, which is largely unmaintained.